File: software/mobile/ios/SVCMobile/Sources/Model/Telemetry.swift
import Foundation

enum MeasurementSource: String, Codable {
    case adc
    case can1
    case imu
    case sensor
    case outputManager = "output-manager"
    case storage
    case firmware
    case derived
}

enum MeasurementQuality: String, Codable {
    case good
    case degraded
    case invalid
    case unavailable
}

struct Measurement<Value: Codable & Equatable>: Codable, Equatable {
    let value: Value?
    let unit: String
    let timestamp: String
    let valid: Bool
    let stale: Bool
    let source: MeasurementSource
    let quality: MeasurementQuality

    var isUsable: Bool {
        valid && !stale && quality != .invalid && quality != .unavailable
    }
}

struct IdentifiedNumberMeasurement: Codable, Equatable {
    let id: String
    let value: Double?
    let unit: String
    let timestamp: String
    let valid: Bool
    let stale: Bool
    let source: MeasurementSource
    let quality: MeasurementQuality
}

struct ChannelTelemetry: Codable, Equatable, Identifiable {
    let id: String
    let current: Measurement<Double>
    let state: Measurement<String>
    let fault: Measurement<String>
}

struct VectorTelemetry: Codable, Equatable {
    let x: Measurement<Double>
    let y: Measurement<Double>
    let z: Measurement<Double>
}

struct VehicleTelemetry: Codable, Equatable {
    let speed: Measurement<Double>
    let engineRpm: Measurement<Double>
    let engineTemperature: Measurement<Double>
    let instantFuelConsumption: Measurement<Double>
    let averageFuelConsumption: Measurement<Double>
    let fuelLevel: Measurement<Double>
    let ambientTemperature: Measurement<Double>
}

struct CANStatus: Codable, Equatable {
    let state: Measurement<String>
    let rxFrames: Measurement<Double>
    let droppedFrames: Measurement<Double>
    let lastFrameTimestamp: Measurement<String>
}

struct StorageStatus: Codable, Equatable {
    let sdCardState: Measurement<String>
    let canLoggerState: Measurement<String>
    let freeBytes: Measurement<Double>
}

struct FirmwareVersions: Codable, Equatable {
    let stm32: Measurement<String>
    let e73: Measurement<String>
    let protocol: Measurement<String>
}

struct DeviceWarning: Codable, Equatable, Identifiable {
    let code: String
    let severity: String
    let message: String
    let timestamp: String
    let active: Bool

    var id: String { code }
}

struct TelemetrySnapshot: Codable, Equatable {
    let schemaVersion: Int
    let protocolVersion: Int
    let deviceId: String
    let sequence: Int
    let timestamp: String
    let batteryVoltage: Measurement<Double>
    let totalCurrent: Measurement<Double>
    let channels: [ChannelTelemetry]
    let powerZoneTemperatures: [IdentifiedNumberMeasurement]
    let ambientLight: Measurement<Double>
    let accelerometer: VectorTelemetry
    let gyroscope: VectorTelemetry
    let leanAngle: Measurement<Double>
    let vehicle: VehicleTelemetry
    let can1: CANStatus
    let storage: StorageStatus
    let versions: FirmwareVersions
    let warnings: [DeviceWarning]
}

File: software/mobile/android/core-model/src/main/kotlin/com/avlyubimov/svc/core/model/Telemetry.kt
package com.avlyubimov.svc.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
enum class MeasurementSource {
    @SerialName("adc") ADC,
    @SerialName("can1") CAN1,
    @SerialName("imu") IMU,
    @SerialName("sensor") SENSOR,
    @SerialName("output-manager") OUTPUT_MANAGER,
    @SerialName("storage") STORAGE,
    @SerialName("firmware") FIRMWARE,
    @SerialName("derived") DERIVED,
}

@Serializable
enum class MeasurementQuality {
    @SerialName("good") GOOD,
    @SerialName("degraded") DEGRADED,
    @SerialName("invalid") INVALID,
    @SerialName("unavailable") UNAVAILABLE,
}

@Serializable
data class Measurement<T>(
    val value: T?,
    val unit: String,
    val timestamp: String,
    val valid: Boolean,
    val stale: Boolean,
    val source: MeasurementSource,
    val quality: MeasurementQuality,
) {
    val isUsable: Boolean
        get() = valid && !stale &&
            quality != MeasurementQuality.INVALID &&
            quality != MeasurementQuality.UNAVAILABLE
}

@Serializable
data class IdentifiedNumberMeasurement(
    val id: String,
    val value: Double?,
    val unit: String,
    val timestamp: String,
    val valid: Boolean,
    val stale: Boolean,
    val source: MeasurementSource,
    val quality: MeasurementQuality,
)

@Serializable
data class ChannelTelemetry(
    val id: String,
    val current: Measurement<Double>,
    val state: Measurement<String>,
    val fault: Measurement<String>,
)

@Serializable
data class VectorTelemetry(
    val x: Measurement<Double>,
    val y: Measurement<Double>,
    val z: Measurement<Double>,
)

@Serializable
data class VehicleTelemetry(
    val speed: Measurement<Double>,
    val engineRpm: Measurement<Double>,
    val engineTemperature: Measurement<Double>,
    val instantFuelConsumption: Measurement<Double>,
    val averageFuelConsumption: Measurement<Double>,
    val fuelLevel: Measurement<Double>,
    val ambientTemperature: Measurement<Double>,
)

@Serializable
data class CanStatus(
    val state: Measurement<String>,
    val rxFrames: Measurement<Double>,
    val droppedFrames: Measurement<Double>,
    val lastFrameTimestamp: Measurement<String>,
)

@Serializable
data class StorageStatus(
    val sdCardState: Measurement<String>,
    val canLoggerState: Measurement<String>,
    val freeBytes: Measurement<Double>,
)

@Serializable
data class FirmwareVersions(
    val stm32: Measurement<String>,
    val e73: Measurement<String>,
    val protocol: Measurement<String>,
)

@Serializable
data class DeviceWarning(
    val code: String,
    val severity: String,
    val message: String,
    val timestamp: String,
    val active: Boolean,
)

@Serializable
data class TelemetrySnapshot(
    val schemaVersion: Int,
    val protocolVersion: Int,
    val deviceId: String,
    val sequence: Long,
    val timestamp: String,
    val batteryVoltage: Measurement<Double>,
    val totalCurrent: Measurement<Double>,
    val channels: List<ChannelTelemetry>,
    val powerZoneTemperatures: List<IdentifiedNumberMeasurement>,
    val ambientLight: Measurement<Double>,
    val accelerometer: VectorTelemetry,
    val gyroscope: VectorTelemetry,
    val leanAngle: Measurement<Double>,
    val vehicle: VehicleTelemetry,
    val can1: CanStatus,
    val storage: StorageStatus,
    val versions: FirmwareVersions,
    val warnings: List<DeviceWarning>,
)
